android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/ListViewProxy.java

```diff
@@ -1,311 +1,499 @@
+
+		// Update table if being re-used.
+		update();
+
+		return view;
 	}
 
-	public void setMarkerHelper(Object marker)
+	/**
+	 * Determine if `marker` event should be fired for specified item.
+	 *
+	 * @param item Item to handle for `marker` event.
+	 */
+	public void handleMarker(ListItemProxy item)
 	{
-		if (marker instanceof HashMap) {
-			HashMap<String, Integer> m = (HashMap<String, Integer>) marker;
-			TiUIView listView = peekView();
-			if (listView != null) {
-				((TiListView) listView).setMarker(m);
-			} else {
-				preloadMarkers.clear();
-				preloadMarkers.add(m);
-			}
+		if (markers.contains(item)) {
+			final int itemIndex = item.indexInSection;
+			final int sectionIndex = this.sections.indexOf(item.getParent());
+			final KrollDict data = new KrollDict();
+
+			// Create and fire marker event.
+			data.put(TiC.PROPERTY_SECTION_INDEX, sectionIndex);
+			data.put(TiC.PROPERTY_ITEM_INDEX, itemIndex);
+			fireEvent(TiC.EVENT_MARKER, data, false);
+
+			// One time event, remove marker.
+			markers.remove(item);
 		}
 	}
 
+	/**
+	 * Insert sections at specified index.
+	 *
+	 * @param index Index to insert sections at.
+	 * @param sections Sections to insert into list.
+	 * @param animation Ignored, for iOS parameter compatibility.
+	 */
 	@Kroll.method
-	public void addMarker(Object marker)
+	public void insertSectionAt(int index, Object sections,
+								@Kroll.argument(optional = true) KrollDict animation)
 	{
-		if (marker instanceof HashMap) {
-			HashMap<String, Integer> m = (HashMap<String, Integer>) marker;
-			TiUIView listView = peekView();
-			if (listView != null) {
-				((TiListView) listView).addMarker(m);
-			} else {
-				preloadMarkers.add(m);
+		final int rawIndex = this.sections.indexOf(getSectionByIndex(index));
+
+		if (rawIndex > -1) {
+			if (sections instanceof Object[]) {
+
+				// Insert ListSection array.
+				for (final Object o : (Object[]) sections) {
+					if (o instanceof ListSectionProxy) {
+						final ListSectionProxy section = (ListSectionProxy) o;
+
+						// Inset ListSection.
+						section.setParent(this);
+						this.sections.add(rawIndex, section);
+					}
+				}
+
+				// Notify ListView of new sections.
+				update();
+
+			} else if (sections instanceof ListSectionProxy) {
+				final ListSectionProxy section = (ListSectionProxy) sections;
+
+				// Insert ListSection.
+				section.setParent(this);
+				this.sections.add(rawIndex, section);
+
+				// Notify ListView of new section.
+				update();
 			}
 		}
 	}
 
-	@Kroll.method
-	public void appendSection(Object section)
+	/**
+	 * Release all views and items.
+	 */
+	@Override
+	public void release()
 	{
-		TiUIView listView = peekView();
-		if (listView != null) {
-			((TiListView) listView).appendSection(section);
-		} else {
-			preload = true;
-			addPreloadSections(section, -1, false);
+		releaseViews();
+
+		if (this.sections != null) {
+			this.sections.clear();
+			this.sections = null;
```

Comment:
I don't think we should null-out the `sections` member variable since the rest of the code assumes it's never null, which is fine. I think the `clear()` method call is enough.

Reply:
In the `release` method I like to `null` out all references to all objects created by the class. We do this in `KrollProxy` and a few others, but you are right that it isn't really necessary. I don't think there's any harm in doing this, as calling `release` assumes the object will never be used again.